source/defender_demo.c

```diff
@@ -0,0 +1,916 @@
+
+    /* Populate device metrics. */
+    if( eMetricsCollectorStatus == eMetricsCollectorSuccess )
+    {
+        xStatus = true;
+        xDeviceMetrics.pxNetworkStats = &( xNetworkStats );
+        xDeviceMetrics.pusOpenTcpPortsArray = &( pusOpenTcpPorts[ 0 ] );
+        xDeviceMetrics.ulOpenTcpPortsArrayLength = ulNumOpenTcpPorts;
+        xDeviceMetrics.pusOpenUdpPortsArray = &( pusOpenUdpPorts[ 0 ] );
+        xDeviceMetrics.ulOpenUdpPortsArrayLength = ulNumOpenUdpPorts;
+        xDeviceMetrics.pxEstablishedConnectionsArray = &( pxEstablishedConnections[ 0 ] );
+        xDeviceMetrics.ulEstablishedConnectionsArrayLength = ulNumEstablishedConnections;
+        xDeviceMetrics.ulStackHighWaterMark = pxTaskStatus.usStackHighWaterMark;
+        xDeviceMetrics.pulTaskIdsArray = pllCustomMetricsTaskNumbers;
+        xDeviceMetrics.ulTaskIdsArrayLength = uxTasksWritten;
+    }
+
+    return xStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvGenerateDeviceMetricsReport( uint32_t * pulOutReportLength )
+{
+    bool xStatus = false;
+    eReportBuilderStatus eReportBuilderStatus;
+
+    /* Generate the metrics report in the format expected by the AWS IoT Device
+     * Defender Service. */
+    eReportBuilderStatus = eGenerateJsonReport( &( pcDeviceMetricsJsonReport[ 0 ] ),
+                                                defenderexampleDEVICE_METRICS_REPORT_BUFFER_SIZE,
+                                                &( xDeviceMetrics ),
+                                                defenderexampleDEVICE_METRICS_REPORT_MAJOR_VERSION,
+                                                defenderexampleDEVICE_METRICS_REPORT_MINOR_VERSION,
+                                                ulReportId,
+                                                pulOutReportLength );
+
+    if( eReportBuilderStatus != eReportBuilderSuccess )
+    {
+        LogError( ( "GenerateJsonReport failed. Status: %d.",
+                    eReportBuilderStatus ) );
+    }
+    else
+    {
+        LogDebug( ( "Generated Report: %.*s.",
+                    *pulOutReportLength,
+                    &( pcDeviceMetricsJsonReport[ 0 ] ) ) );
+        xStatus = true;
+    }
+
+    return xStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvPublishDeviceMetricsReport( uint32_t reportLength )
+{
+    static MQTTPublishInfo_t xPublishInfo = { 0 };
+    CommandInfo_t xCommandParams = { 0 };
+    MQTTStatus_t xCommandAdded;
+
+    xPublishInfo.qos = MQTTQoS1;
+    xPublishInfo.pTopicName = DEFENDER_API_JSON_PUBLISH( democonfigCLIENT_IDENTIFIER );
+    xPublishInfo.topicNameLength = DEFENDER_API_LENGTH_JSON_PUBLISH( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xPublishInfo.pPayload = &( pcDeviceMetricsJsonReport[ 0 ] );
+    xPublishInfo.payloadLength = reportLength;
+
+    xCommandParams.blockTimeMs = defenderexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = NULL;
```

Comment:
I'd consider adding a comment here that we don't need the completion callback since we expect to get a response on the accepted/rejected topics